### DISCUSSION TITLE
How to map  a relational database

### DISCUSSION THREAD

**Discussion by hyzx86** [OrchardCMS/OrchardCore#9911] [Q&A] [answered]
We have a relational database whose data is extracted from other systems through ETL.
Some of the data is imported manually, and some are generated by other systems

I hope to realize their unified management and query through OC , such as:

1. Field mapping, can generate a Content Type by selecting a table. In most cases, you can use the default fields. I plan to use a management page to achieve this function

2. In the first step, it should be possible to automatically add a `Content Item` to the pre-existing `Content Type`, which is used to save the configuration information of how to process the mapping

3. Use the existing mapping relationship to realize CRUD , import and export

4. Realize two-way refresh of data

I noticed that the workflow can monitor type changes, and I can develop an Activity to handle it. But if I process a batch import, it should trigger a lot of data updates

Do you have any better suggestions for this case? 

Thanks!

**Comment by aaronamm** [comment, marked as answer]
@hyzx86 I am not sure you have checked my my [PoC project](https://github.com/codesanook/OrchardCore.EF). 
It enable EF for a non content part model and contains a central DbContext that other modules can use.

However, it look like my project can help you much on this. I haven't done with import existing data to a content item in Orchard Core CMS.
I think you may need for a while until someone give an idea if there is a native way to import data to Orchard Core CMS.

However, here is what I can think of.

- You many need to define  content part, content type to match your existing model.
- Then, convert your data to JSON of content item and save to Document and ContentItemIndex.
- You need to figure it out how to generate id. 
- I wonder if you can have a command line tool to insert data and use some Orchard Core packages.

**Reply by hyzx86** [reply]
If using the command line makes things complicated, 
I think it's better to use a page tool to generate structural mappings

**Reply by Skrypt** [reply]
Suggestion is to use the ContentManager class to do this. Or generate a recipe.json file that will be imported through the import / export module classes/methods which support bulk import of data.

**Reply by hyzx86** [reply]
Hi @Skrypt  Do you have any suggestions on how to obtain the following data?

I hope to get its corresponding simple Json description from the existing contentItem:
```
"MyType": {
    "Field1": {
      "Text": "TestTable"
    },
    "Field2": {
      "Value": "TestContentValue"
    },
    "Field3": {
      "Text": "{\"JsonData\":\"123123\"}"
    }
  }
```

convert to：

```
{
    "MyType": {
        "Field1": "TestTable",
        "Field2": "TestContentValue",
        "Field3": "{\"JsonData\":\"123123\"}"
    }
}
```

Or I can only check the types one by one and judge

**Reply by hyzx86** [reply]
found in class `ContentFieldsProvider `

**Reply by Skrypt** [reply]
I'm not sure why you want to change the JSON schema of these Fields. Changing the structure of the JSON will only lead you to incompatibilities moving forward. If you need to import data then you should stick to the JSON schema that we use.

Your question is really specific but we don't know the context so it is hard to advise properly.

**Reply by hyzx86** [reply]
I need to get the JSON Path of the main fields from the different contentTypes to generate the mapping to the relational database
just like this:
![image](https://user-images.githubusercontent.com/15613121/126688852-a3c0e418-e474-48a6-bf01-6c783c5ae242.png)
